go/parquet/file/column_writer.go

```diff
@@ -0,0 +1,567 @@
+	chunkStatistics metadata.TypedStatistics
+
+	// total number of values stored in the data page. this is the maximum
+	// of the number of encoded def levels or encoded values. for
+	// non-repeated, required columns, this is equal to the number of encoded
+	// values. For repeated or optional values, there may be fewer data values
+	// than levels, and this tells you how many encoded levels there are in that case
+	numBuffered int64
+
+	// the total number of stored values. for repeated or optional values. this
+	// number may be lower than numBuffered
+	numBufferedEncoded int64
+
+	rowsWritten       int
+	totalBytesWritten int64
+	// records the current number of compressed bytes in a column
+	totalCompressedBytes int64
+	closed               bool
+	fallback             bool
+
+	pages []DataPage
+
+	defLevelSink *encoding.PooledBufferWriter
+	repLevelSink *encoding.PooledBufferWriter
+
+	uncompressedData bytes.Buffer
+	compressedTemp   *bytes.Buffer
+
+	currentEncoder encoding.TypedEncoder
+}
+
+func newColumnWriterBase(metaData *metadata.ColumnChunkMetaDataBuilder, pager PageWriter, useDict bool, enc parquet.Encoding, props *parquet.WriterProperties) columnWriter {
+	ret := columnWriter{
+		metaData:     metaData,
+		descr:        metaData.Descr(),
+		levelInfo:    computeLevelInfo(metaData.Descr()),
+		pager:        pager,
+		hasDict:      useDict,
+		encoding:     enc,
+		props:        props,
+		mem:          props.Allocator(),
+		defLevelSink: encoding.NewPooledBufferWriter(0),
+		repLevelSink: encoding.NewPooledBufferWriter(0),
+	}
+	if pager.HasCompressor() {
+		ret.compressedTemp = new(bytes.Buffer)
+	}
+	if props.StatisticsEnabledFor(ret.descr.Path()) && ret.descr.SortOrder() != schema.SortUNKNOWN {
+		ret.pageStatistics = metadata.NewStatistics(ret.descr, props.Allocator())
+		ret.chunkStatistics = metadata.NewStatistics(ret.descr, props.Allocator())
+	}
+
+	if ret.props.DataPageVersion() == parquet.DataPageV1 {
+		if ret.descr.MaxDefinitionLevel() > 0 {
+			ret.defLevelSink.SetOffset(arrow.Uint32SizeBytes)
+		}
+		if ret.descr.MaxRepetitionLevel() > 0 {
+			ret.repLevelSink.SetOffset(arrow.Uint32SizeBytes)
+		}
+	}
+
+	ret.defEncoder.Init(parquet.Encodings.RLE, ret.descr.MaxDefinitionLevel(), ret.defLevelSink)
+	ret.repEncoder.Init(parquet.Encodings.RLE, ret.descr.MaxRepetitionLevel(), ret.repLevelSink)
+	return ret
+}
+
+func (w *columnWriter) SetBitsBuffer(buf *memory.Buffer) { w.bitsBuffer = buf }
+
+func (w *columnWriter) LevelInfo() LevelInfo { return w.levelInfo }
+
+func (w *columnWriter) Type() parquet.Type {
+	return w.descr.PhysicalType()
+}
+
+func (w *columnWriter) Descr() *schema.Column {
+	return w.descr
+}
+
+func (w *columnWriter) Properties() *parquet.WriterProperties {
+	return w.props
+}
+
+func (w *columnWriter) TotalCompressedBytes() int64 {
+	return w.totalCompressedBytes
+}
+
+func (w *columnWriter) TotalBytesWritten() int64 {
+	return w.totalBytesWritten
+}
+
+func (w *columnWriter) RowsWritten() int64 {
```

Comment:
is there a strong reason to up-cast this to int64?

Reply:
I was being consistent with the other functions (returning numbers of bytes) but to be fair, it kinda makes sense to just leave this as an `int`.